fix(forms) : Broken link NG_VALIDATORS replace by an example

wKoza opened the pull request:
**Please check if the PR fulfills these requirements**
- [X ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ X ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
Broken link in NG_VALIDATORS jsdoc
see https://angular.io/docs/ts/latest/api/forms/index/NG_VALIDATORS-let.html

**What is the new behavior?**
replace link by a use case


**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ X ] No
```

wKoza committed:
fix(forms) : Break link NG_VALIDATORS replace by an example

wKoza committed:
fix(forms) : Break link NG_VALIDATORS replace by an example

tytskyi commented on modules/@angular/forms/src/validators.ts:
````diff
@@ -27,7 +27,18 @@ function isEmptyInputValue(value: any): boolean {
+ * ```typescript
+ * @Directive({
+ *   selector: '[custom-validator]',
+ *   providers: [{provide: NG_VALIDATORS, useExisting: CustomValidatorDirective, multi: true}]
````
forgotten `forwardRef`?

DzmitryShylovich replied:
not required here

Commit:
fix(core): stringify shouldn't throw when toString returns null/undefined (#14975)

Fixes #14948

PR Close #14975

gkalpak committed:
docs(compiler-cli): mention that `.ngsummary.json` files should be gitignore'd (#15047)

PR Close #15047

alxhub committed:
fix(platform-server): interpret Native view encapsulation as Emulated on the server (#15155)

PR Close #15155

mhevery committed:
fix(core): update peer dep on zone.js to ^0.8.5 (#15365)

Closes #15185

PR Close #15365

IgorMinar committed:
build: update to angular 4.0.0-rc.4 and service-worker@1.0.0-beta.7

petebacondarwin committed:
build(aio): update to dgeni-packages 0.17.0

This new version fixes the problem with mistaking
content inside inline HTML blocks as tags

Related to #15199

petebacondarwin committed:
build(aio): fix generated github links

The change from `modules` to `packages` needed to be applied to these links.

wardbell committed:
feat(aio): expand nav menu branch of select item on page load

petebacondarwin committed:
ci(aio): lower prerender browser instances

Hopefully this will reduce aio test flakes

kara committed:
ci(forms): add tina as secondary for forms (#15262)

PR Close #15262

petebacondarwin committed:
build(aio): do not render private classes and members

petebacondarwin committed:
build(aio): do not render ignored docs

petebacondarwin committed:
build(aio): use forked version of Rho rendering engine

The original Rho is too strict when it comes to markdown headings.
It requires that there be a blank line separating the heading and the
next paragraph.  The forked version here fixes that; but the Rho project
will not merge it as it goes against there basic rules.

petebacondarwin committed:
build(aio): ensure that internal document links work with base href

petebacondarwin committed:
build(aio): remove cheatsheet processing and content

This will be replaced by a single file migrated from angular.io

petebacondarwin committed:
build(aio): don't process unnecessary example files in doc-gen

This lowers the `yarn docs` processing time by about 40% (from 50 secs
to 30 secs).

sis0k0 committed:
fix(animations): correct the main entry path in package.json (#15300)

PR Close #15300

ocombe committed:
fix(compiler): use attribute id to merge translations (#15302)

We extracted ids from i18n attributes but forgot to use them when merging the translations, resulting in an error about missing translations even when they were correctly defined.

Fixes #15234

PR Close #15302

mhevery committed:
fix: remove left over debugger statement (#15309)

PR Close #15309

matsko committed:
fix(animations): stringify boolean values as `1` and `0` (#15311)

Closes #15247
Closes #15311

PR Close #15311

matsko committed:
fix(animations): ensure enter/leave cancellations work (#15323)

Closes #15315
Closes #15323

PR Close #15323

mhevery committed:
Revert "fix(compiler): shouldn't throw when Symbol is used as DI token (#13701)" (#15319)

This reverts commit 8b5c6b2732f03927b5b697d22c5d84e8fab7e225.

This feature is not compatible with the `Injector.get` which now only 
takes `Type` or `InjectableToken`. `Symbol` is not a valid type.

Closes #15183

PR Close #15319

petebacondarwin committed:
build(aio): add version into navigation.json

The navigation.json is now passed through the dgeni pipeline.
The source file has been moved to `aio/content/navigation.json`
but the generated file will now appear where the original source file
was found, `aio/src/content/navigation.json`.

Everything inside `aio/src/content` is now generated and ignored by git.

The `processNavigationMap` processor in this commit adds the current version
information to the navigation.json file and verifies the relative urls in
the file map to real documents.

The navigationService exposes the versionInfo as an observable, which the
AppComponent renders at the top of the sidenav.

tbosch committed:
fix(compiler): don’t call `check` if we don’t need to (#15322)

If a directive has not bindings nor has a `ngDoCheck` / `ngOnInit`
lifecycle hook, don’t generate a `check` call.

This does not have an impact on the behavior, but produces
less code.

PR Close #15322

chuckjaz committed:
feat(tsc-wrapped): record original location of flattened symbols (#15367)

Added an "origins" section to the flat module `.metadata.json` files
that records where the original symbols was declared. This allows
correctly calculating relative path references recorded in metadata.

chuckjaz committed:
fix(compiler): look for flat module resources using declaration module path (#15367)

`ngc` would look for flat module resources relative to the flat module index.
`ngc` now looks for flat module resources relative to the `.d.ts` file that declarates
the component.

Fixes #15221

PR Close #15367

matsko committed:
fix(animations): ensure empty animate() steps work at the end of a sequence (#15328)

Closes #15310
Closes #15328

PR Close #15328

kara committed:
fix(forms): make composition event buffering configurable (#15256)

This commit fixes a regression where `ngModel` no longer syncs
letter by letter on Android devices, and instead syncs at the
end of every word. This broke when we introduced buffering of
IME events so IMEs like Pinyin keyboards or Katakana keyboards
wouldn't display composition strings. Unfortunately, iOS devices
and Android devices have opposite event behavior. Whereas iOS
devices fire composition events for IME keyboards only, Android
fires composition events for Latin-language keyboards. For
this reason, languages like English don't work as expected on
Android if we always buffer. So to support both platforms,
composition string buffering will only be turned on by default
for non-Android devices.

However, we have also added a `COMPOSITION_BUFFER_MODE` token
to make this configurable by the application. In some cases, apps
might might still want to receive intermediate values. For example,
some inputs begin searching based on Latin letters before a
character selection is made.

As a provider, this is fairly flexible. If you want to turn
composition buffering off, simply provide the token at the top
level:

```ts
providers: [
   {provide: COMPOSITION_BUFFER_MODE, useValue: false}
]
```

Or, if you want to change the mode  based on locale or platform,
you can use a factory:

```ts
import {shouldUseBuffering} from 'my/lib';

....
providers: [
   {provide: COMPOSITION_BUFFER_MODE, useFactory: shouldUseBuffering}
]
```

Closes #15079.

PR Close #15256

chuckjaz committed:
build: remove use of alias metadata (#15331)

chuckjaz committed:
revert: feat(compiler-cli): support metadata file aliases (#15331)

This reverts commit 0ab49d4cec23e2957fd7e8dad55d15fc539c69ed.

PR Close #15331

IgorMinar committed:
build(aio): update @angular to 4.0.0-rc.5 and zone.js to 0.8.4

IgorMinar committed:
build(aio): upgrade @angular/cli to 1.0.0-rc.4

petebacondarwin committed:
test(aio): use MockLogger in AppComponent tests

This component sends a lot of messages to the console, which makes the
test run less easy to follow.

sjtrimble committed:
feat(aio): font change and toolbar shadow

petebacondarwin committed:
build(aio): fix paths to "index" pages

Content pages like `tutorial/index.md` were being mapped to `tutorial.index.json`,
which meant that they could only be rendered if you browsed to `/tutorial/index`.

This didn't sit well so now these pages are mapped to `tutorial.json`, which
means that you browser to them via `/tutorial/` or just `/tutorial`.

Fixed #15335

sjtrimble committed:
refactor(aio): removed dead code

petebacondarwin committed:
test(aio): fix test descriptions for ordering of processors

petebacondarwin committed:
build(aio): ensure that internal document links work with base href

petebacondarwin committed:
build(aio): add version into navigation.json

The navigation.json is now passed through the dgeni pipeline.
The source file has been moved to `aio/content/navigation.json`
but the generated file will now appear where the original source file
was found, `aio/src/content/navigation.json`.

Everything inside `aio/src/content` is now generated and ignored by git.

The `processNavigationMap` processor in this commit adds the current version
information to the navigation.json file and verifies the relative urls in
the file map to real documents.

The navigationService exposes the versionInfo as an observable, which the
AppComponent renders at the top of the sidenav.

gkalpak committed:
revert: build(aio): implement prerendering (#15346)

This reverts commit d0bc83ca275964fa1de6684c3cdc6714de386efd.

Protractor-based prerendering is flakey on Travis and takes several minutes to
complete, slowing down the build. Prerendering has a lower impact now that we
use a ServiceWorker. We will revisit in the future (probably using a
`PlatformServer`-based approach).

PR Close #15346

tbosch committed:
fix(core): provide `NgModuleRef` in `ViewContainerRef.createComponent`. (#15350)

This is needed to support the corner cases:
- usage of a `ComponentFactory` that was created on the fly via `Compiler`
- overwriting of the `NgModuleRef` that is associated to a
  `ComponentFactory` by the `ComponentFactoryResolver` from
  which it was read.

Fixes #15241

IgorMinar committed:
refactor(aio): cleanup polyfill imports and realign them with the latest cli blueprint

This shouldn't change anything. But it's interesting that we used to have this  import
that seemed bogus, but there were no compilation or rutime errors.

IgorMinar committed:
build(aio): upgrade to rxjs@5.2.0

tbosch committed:
refactor(compiler): only produce `log` expressions for elements / text (#15350)

This change reduces the amount of generated code by only adding `log`
calls for elements and text nodes.

We need the `log` calls to allow users to jump to the right place
in the template via source maps. However, we only need it for element
and text nodes, but not for directives, queries, … as for them we 
first locate the corresponding element or text node.

Related to #15239

PR Close #15350

alxhub committed:
fix(platform-server): throw a better error message for relative URLs (#15357)

Unlike in the browser, on the server there is no concept of a document origin.
Thus, it is illegal to make requests for relative URLs against Http on platform-server.

Currently this fails with a vague error:

Error: Uncaught (in promise): Error at resolvePromise

This change adds explicit validation and a friendlier error message:

Error: URLs requested via Http on the server must be absolute. URL: /testing

Another option considered was to track the concept of an origin for the platform
and automatically prepend it to relative URLs. This would cause automatic "local
RPCs" to be made, though, which would be an unexpected and undesirable default
behavior.

Fixes #15349

PR Close #15357

tbosch committed:
fix(core): mark components for check when host events trigger. (#15359)

Fixes #15352

PR Close #15359

vicb committed:
refactor(core): misc cleanup (#15366)

PR Close #15366

chuckjaz committed:
fix(tsc-wrapped): use windows friendly path normalization in bundler (#15374)

Fixes #15289

PR Close #15374

petebacondarwin committed:
build(aio): output `{@example}` tags as `<code-example>` elements (#15382)

kuncevic committed:
docs: spelling errors (#15375)

wKoza committed:
docs: update 4.0.0-rc.5 release schedule (#15288)

ocombe committed:
fix(compiler-cli): adding missing format xliff for the extractor (#15386)

To generate XLF files with ng-xi18n we could use the format parameter "xlf" or "xlif". The real name is "xliff" not "xlif", so this probably was a typo. This PR adds "xliff" as can be expected

vicb committed:
fix(upgrade): component injectors should not link the module injector tree (#15385)

gkalpak committed:
fix(aio): do not fallback to `index.html` for file requests (#15401)

Previously, all URLs were rewritten to `index.html` in order to support
deep-linking. This works when navigating to URLs that correspond to existing
resources. E.g. navigating to `/tutorial` returns `index.html` and then the
`DocViewer` takes over and requests `tutorial.json`.
Navigating to a non-existent URL (e.g. `/foo`), will return `index.html`, which
in turn requests (the non-existent) `foo.json` and throws an error when trying
to parse the returned `index.html` as JSON.

This commit fixes it by only rewriting URLs that do not request a file (i.e. do
not include a `.` in the last path segment).

Fixes #15398

jnizet committed:
docs(router): fix typo in ParamMap api doc (#15397)

kuncevic committed:
docs: spelling error (#15406)

alexeagle committed:
ci(travis): update excluded branch to g3 (#15391)

jsayol committed:
docs(common): update ngFor docs to new `as` syntax (#15166)

kuncevic committed:
docs: fix spelling

kuncevic committed:
docs: spelling errors

robwormald committed:
fix(compiler): only log template deprecation warning once (#15364)

IgorMinar committed:
docs: move markdown docs from root dir to docs/ and remove obsolete files (#15410)

IgorMinar committed:
release: cut the 4.0.0-rc.6 release

IgorMinar committed:
build(aio): upgrade to @angular@4.0.0-rc.6 (#15412)

petebacondarwin committed:
build(aio): delete content folder before doc-gen (#15414)

This should prevent false-positive e2e test runs where stale files
are left lying around.

Commit:
fix(router): should pass new data to Observable when query params change (#15387)

Fixes #15290

Commit:
fix(platform-browser): setAttribute should work with xmlns namespace (#14874)

Closes #14865

tbosch committed:
fix(compiler): assume queries with no matches as static (#15429)

This has the side effect of allowing `@Input` and `@ContentChild`
on the same property if the query is static (see the bug description
for details).

Fixes #15417

chuckjaz committed:
fix(compiler): correctly handle when `toString` is exported (#15430)

Fixes #15420

IgorMinar committed:
fix: prevent strictNullChecks support until #15432 is fixed (#15434)

vicb committed:
release: cut the 4.0.0 release

vicb committed:
docs: add changelog for 4.0.0

IgorMinar committed:
docs: add a note about exclusion of symbols prefixed with ɵ from our public api surface (#15440)

IgorMinar committed:
docs: add the 4.0.0 code name to the changelog

IgorMinar committed:
docs: update CONTRIBUTING link in README.md

IgorMinar committed:
docs: update the contributing link one more time :-)

IgorMinar committed:
docs: revert the move of CONTRIBUTING.md to docs/CONTRIBUTING.md

Because the root path has a special meeting for GitHub, Jeremy, unicorns and red arrows.

thomaspink committed:
docs: fixed broken links (#15455)

Closes #15446

wKoza committed:
Merge branch 'fix_path_NG_VALIDATORS' of https://github.com/wKoza/angular into fix_path_NG_VALIDATORS

# Conflicts:
#	modules/@angular/forms/src/validators.ts
